src/daft-csv/src/deserialize.rs

```diff
@@ -0,0 +1,304 @@
+}
+
+/// Deserializes bytes to a single i128 representing a decimal
+/// The decimal precision and scale are not checked.
+#[inline]
+fn deserialize_decimal(bytes: &[u8], precision: usize, scale: usize) -> Option<i128> {
+    let mut a = bytes.split(|x| *x == b'.');
+    let lhs = a.next();
+    let rhs = a.next();
+    match (lhs, rhs) {
+        (Some(lhs), Some(rhs)) => lexical_core::parse::<i128>(lhs).ok().and_then(|x| {
+            lexical_core::parse::<i128>(rhs)
+                .ok()
+                .map(|y| (x, lhs, y, rhs))
+                .and_then(|(lhs, lhs_b, rhs, rhs_b)| {
+                    let lhs_s = significant_bytes(lhs_b);
+                    let rhs_s = significant_bytes(rhs_b);
+                    if lhs_s + rhs_s > precision || rhs_s > scale {
+                        None
+                    } else {
+                        Some((lhs, rhs, rhs_s))
+                    }
+                })
+                .map(|(lhs, rhs, rhs_s)| lhs * 10i128.pow(rhs_s as u32) + rhs)
+        }),
+        (None, Some(rhs)) => {
+            if rhs.len() != precision || rhs.len() != scale {
+                return None;
+            }
+            lexical_core::parse::<i128>(rhs).ok()
+        }
+        (Some(lhs), None) => {
+            if lhs.len() != precision || scale != 0 {
+                return None;
+            }
+            lexical_core::parse::<i128>(lhs).ok()
+        }
+        (None, None) => None,
+    }
+}
+
+#[inline]
+fn deserialize_boolean<B, F>(rows: &[B], column: usize, op: F) -> Box<dyn Array>
+where
+    B: ByteRecordGeneric,
+    F: Fn(&[u8]) -> Option<bool>,
+{
+    let iter = rows.iter().map(|row| match row.get(column) {
+        Some(bytes) => {
+            if bytes.is_empty() {
+                return None;
+            }
+            op(bytes)
+        }
+        None => None,
+    });
+    Box::new(BooleanArray::from_trusted_len_iter(iter))
+}
+
+#[inline]
+fn deserialize_utf8<O: Offset, B: ByteRecordGeneric>(rows: &[B], column: usize) -> Box<dyn Array> {
+    let iter = rows.iter().map(|row| match row.get(column) {
+        Some(bytes) => to_utf8(bytes),
+        None => None,
+    });
+    Box::new(Utf8Array::<O>::from_trusted_len_iter(iter))
+}
+
+#[inline]
+fn deserialize_binary<O: Offset, B: ByteRecordGeneric>(
+    rows: &[B],
+    column: usize,
+) -> Box<dyn Array> {
+    let iter = rows.iter().map(|row| row.get(column));
+    Box::new(BinaryArray::<O>::from_trusted_len_iter(iter))
+}
+
+#[inline]
+fn deserialize_null<B: ByteRecordGeneric>(rows: &[B], _: usize) -> Box<dyn Array> {
+    // TODO(Clark): Once we add strict parsing mode, where failure to parse to an intended dtype fails instead
+    // of resulting in a null, add an explicit parsing pass over the column.
+    Box::new(NullArray::new(DataType::Null, rows.len()))
+}
+
+#[inline]
+fn deserialize_naive_datetime(string: &str) -> Option<chrono::NaiveDateTime> {
+    for fmt in ALL_NAIVE_TIMESTAMP_FMTS {
```

Comment:
if the string is `RFC3339_WITH_SPACE_NO_TIME_ZONE_NO_FRACTIONAL`, we would iterate through 6 possible choices before choosing it again. Maybe we should also return the `index` of the choice we made and try that first for the next iteration before falling back to the rest of possible choices.